ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java

```diff
@@ -557,80 +556,95 @@ public String toString() {
+      future.whenComplete(
+          (peerHeadBlockHeader, error) -> {
+            // If we successfully retrieved the chain head, check trailing peer requirements and
+            // update
+            // the peer's estimated height.
+            if (peerHeadBlockHeader != null) {
+              checkTrailingPeerRequirements(peer);
+              peer.chainState().updateHeightEstimate(peerHeadBlockHeader.getNumber());
             } else {
-              isServingSnapFuture = CompletableFuture.completedFuture(null);
+              LOG.debug(
+                  "Failed to retrieve chain head info. Disconnecting {}... {}",
+                  peer.getLoggableId(),
+                  error);
+              peer.disconnect(
+                  DisconnectMessage.DisconnectReason.USELESS_PEER_FAILED_TO_RETRIEVE_CHAIN_HEAD);
             }
-            isServingSnapFuture.thenRun(
-                () -> {
-                  if (!peer.getConnection().isDisconnected() && addPeerToEthPeers(peer)) {
-                    connectedPeersCounter.inc();
-                    connectCallbacks.forEach(cb -> cb.onPeerConnected(peer));
-                  }
-                });
+          });
+    }
+
+    // If we are doing snap sync, check if the peer is a snap server
+    checkSnapServer(peer);
+  }
+
+  private void checkTrailingPeerRequirements(final EthPeer peer) {
+    // we can check trailing peers now
+    final TrailingPeerRequirements trailingPeerRequirements =
+        trailingPeerRequirementsSupplier.get();
+    if (trailingPeerRequirements != null) {
+      if (peer.chainState().getEstimatedHeight()
+          < trailingPeerRequirements.getMinimumHeightToBeUpToDate()) {
+        if (!(getNumTrailingPeers(trailingPeerRequirements.getMinimumHeightToBeUpToDate())
+            < trailingPeerRequirements.getMaxTrailingPeers())) {
+          LOG.atTrace()
+              .setMessage(
+                  "Adding trailing peer {} would exceed max trailing peers {}. Disconnecting...")
+              .addArgument(peer.getLoggableId())
+              .addArgument(trailingPeerRequirements.getMaxTrailingPeers())
+              .log();
+          peer.disconnect(DisconnectMessage.DisconnectReason.USELESS_PEER_EXCEEDS_TRAILING_PEERS);
+        }
+      }
+    }
+  }
+
+  private void checkSnapServer(final EthPeer peer) {
+    CompletableFuture<Void> isServingSnapFuture;
+    if (syncMode == SyncMode.SNAP || syncMode == SyncMode.CHECKPOINT) {
+      // even if we have finished the snap sync, we still want to know if the peer is a snap
+      // server
+      isServingSnapFuture =
+          CompletableFuture.runAsync(
+              () -> {
+                try {
+                  checkIsSnapServer(peer);
+                } catch (Exception e) {
+                  throw new RuntimeException(e);
+                }
+              });
+    } else {
+      isServingSnapFuture = CompletableFuture.completedFuture(null);
+    }
+    isServingSnapFuture.thenRun(
```

Comment:
This does not really belong into the check snap server method. Probably return the future and do that in the status exchanged method.